Theme Designer: color picker card + samples card.

aneeshack4 opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #8390 #8389 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Created the layout of the app with cards for all the different sections and made the color picker functional and the samples panel with all the components. 

TODO: 
- [x] fix relative styling of color picker's label with the color square & textfield - having some trouble with ThemeDesignerColorPicker's usage of Stack to make this happen.
- [ ] add Radio Button: there isn't a Fabric control for this but it looks like I might be able to use choiceGroup to make it?

NOTE: Just made this PR to start getting feedback early and on parts - not really meant for merging right away.

looks like this right now:
![image](https://user-images.githubusercontent.com/4161791/54857885-ad3e6700-4ccf-11e9-895f-00cb04049f6c.png)

#### Focus areas to test

Make sure it looks like the [mocks](https://www.figma.com/file/f8c3zkZpY7DihNjpWpiAJfc4/Theme-Designer--%F0%9F%96%8D?node-id=52%3A184)


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8442)

khmakoto commented on apps/theming-designer/src/components/ThemeDesignerColorPicker.tsx:
```diff
@@ -0,0 +1,22 @@
+    props.onChange(newVal);
+  };
+  return (
+    <Stack horizontal gap={10}>
```
If you use `horizontalAlign: stretch` in here I think that would help with alignment with the `ThemeDesignerColorPicker`.

aneeshack4 commented:
I know there's a merge conflict with shrinkwrap.yaml - I tried to resolve with `npm run generate`. That failed because version numbers were off in @uifabric/icons, ts-lint, prettier, uifr so I corrected them & ran `npm run generate` which passed. Then I ran "rush build" to get the shrinkwrap file to update but it failed with these errors in react-cards. A bunch of others seem to have gotten these [errors](https://github.com/facebook/react/issues/14846) too and the common solution seemed to be updating the `@types/react` version but that didn't work here.

ecraig12345 commented on apps/theming-designer/src/components/app.scss:
```diff
@@ -0,0 +1,38 @@
+@import '~office-ui-fabric-react/dist/sass/References';
+
+body,
```
Can you use one of the css-in-js approaches instead? Sass files create a lot of headaches for build and we'd ideally like to get rid of them. 

For what you're doing, `mergeStyleSets` should work well. Here's an example to help you get started (note the special syntax for applying global styles to the `html` element):
```ts
const styles = mergeStyleSets({
  app: {
    display: 'flex',
    selectors: {
      ':global(html)': {
        padding: 0,
        margin: 0
      }
    }
  },
  sidebar: {
    flex: '0 0 400px',
    background: '#fff'
  }
});
```

ecraig12345 commented on apps/theming-designer/src/components/Samples.tsx:
```diff
@@ -0,0 +1,132 @@
+  Text,
+  TextField,
+  Toggle
+} from '../../../../packages/office-ui-fabric-react/lib';
```
Instead of relative path imports, you should be able to use just `office-ui-fabric-react` (or `office-ui-fabric-react/lib/whatever`). If that's causing compiler errors, paste the error here and we'll figure out how to fix it.

aneeshack4 replied:
I tried to do: 
`import {
  ActionButton,
  Checkbox,
  DefaultButton,
  Dropdown,
  Link,
  PrimaryButton,
  Stack,
  Text,
  TextField,
  Toggle
} from 'office-ui-fabric-react/lib';`

but I'm getting the following error and looks like I'm not the only one because when I googled this I found that it happen in [other parts](https://github.com/OfficeDev/office-ui-fabric-react/issues/8146) of the Fabric code too. 
![image](https://user-images.githubusercontent.com/4161791/55022231-61354000-4fb7-11e9-8a45-643111b74ad1.png)

ecraig12345 replied:
Remove the `lib` part and it should work. Either that, or add an `index` segment after `lib`.

ecraig12345 replied:
Actually I take that back...just add an `index` segment.

ecraig12345 commented on common/config/rush/shrinkwrap.yaml:
```diff
@@ -28,6 +28,7 @@ dependencies:
   '@rush-temp/test-bundles': 'file:projects/test-bundles.tgz'
   '@rush-temp/test-utilities': 'file:projects/test-utilities.tgz'
   '@rush-temp/theme-samples': 'file:projects/theme-samples.tgz'
+  '@rush-temp/theming-designer': 'file:projects/theming-designer.tgz'
```
To resolve the shrinkwrap issue, I'd recommend copying the current shrinkwrap file from master into your branch: `git checkout master -- common/config/rush/shrinkwrap.yaml`. *Then* running `rush generate` (or `npm run update`, which does the same thing). This way the amount of new stuff it has to figure out how to generate will be much smaller.

jdhuntington commented on apps/theming-designer/index.html:
```diff
@@ -0,0 +1,16 @@
+
+    <title>Theming Designer App</title>
+    <script type="text/javascript" src="//cdnjs.cloudflare.com/ajax/libs/react/16.3.2/umd/react.production.min.js"></script>
+    <script type="text/javascript" src="//cdnjs.cloudflare.com/ajax/libs/react-dom/16.3.2/umd/react-dom.production.min.js"></script>
```
Would be good to target react 16.8.something.

ecraig12345 replied:
If you want to add 16.8 (and matching types) as a dep for just this project, you'll need to add it to common-versions.json allowedAlternateVersions.

jdhuntington commented on apps/theming-designer/package.json:
```diff
@@ -0,0 +1,38 @@
+  ],
+  "devDependencies": {
+    "@types/es6-promise": "0.0.32",
+    "@types/react": "16.3.16",
```
16.8.+

jdhuntington commented on apps/theming-designer/src/components/ColorPanel.tsx:
```diff
@@ -0,0 +1,64 @@
+          <ThemeDesignerColorPicker
+            label={'Background Color'}
+            value={this.state.backgroundColor}
+            onChange={(v: string) => this.setState({ backgroundColor: v })}
```
best to turn these `onChange` callbacks into member functions in the class, then refer to them directly. Saves React some effort on re-render.

KevinTCoughlin replied:
100% this

jdhuntington commented on apps/theming-designer/src/components/SamplesCardHeader.tsx:
```diff
@@ -0,0 +1,16 @@
+import { Text, ITextProps } from '../../../../packages/office-ui-fabric-react/lib/Text';
+import './SamplesCardHeader.scss';
+
+interface ISamplesCardHeaderProps {
```
Nit: A more intuitive API would probably be to accept `children` instead of `label`. That way, a caller can do this instead:

    <SampleCardHeader>Input</SampleCardHeader>

jdhuntington commented on apps/theming-designer/src/components/SamplesCardHeader.scss:
```diff
@@ -0,0 +1,5 @@
+:global {
+  .samplescardheader {
+    border-bottom: 1px black solid;
```
Might want to add `text-transform: uppercase;` to keep in line with the styling, and not require callers to know that things should be upcased.

KevinTCoughlin commented:
FYI - Draft PRs are helpful for this sort of WIP approach, see https://github.blog/2019-02-14-introducing-draft-pull-requests/. This is because they are not merge-able until promoted from draft status.

KevinTCoughlin commented on apps/theming-designer/index.html:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html dir="ltr">
```
Nit, you may also want to specify `lang="en"` here unless its handled dynamically by our library.

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang

KevinTCoughlin commented on apps/theming-designer/src/components/AccChecker.tsx:
```diff
@@ -0,0 +1,15 @@
+import { Card } from '@uifabric/react-cards';
+//import { ColorsPage } from '@uifabric/fabric-website-resources/lib/components/pages/ColorsPage';
+
+export class AccChecker extends React.Component {
```
Prefer fully qualified names to help with readability for developers less familiar with the functionality.

KevinTCoughlin commented on apps/theming-designer/src/components/ColorPanel.tsx:
```diff
@@ -0,0 +1,64 @@
+}
+
+export class ColorPanel extends React.Component<IColorPanelProps, IColorPanelState> {
+  constructor(props: any) {
```
Try to avoid `any` type as much as possible to help with development and maintainability.

This `any` is `IColorPanelProps`.

Also may also want to consider picking the types for your state interface from props to keep them in-sync. Assuming both interfaces are necessary.

KevinTCoughlin commented on apps/theming-designer/src/components/Palette.tsx:
```diff
@@ -0,0 +1,101 @@
+      <Card styles={{ root: { width: '800px', height: 'auto' } }}>
+        <h1>Fabric Palette</h1>
+        <table>
+          <tr>
```
You may want to render a `thead`, `tbody`, etc. to be more semantically correct. Refer to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table

KevinTCoughlin commented on apps/theming-designer/src/components/Palette.tsx:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  private _slotWidget = (slotRule: IThemeSlotRule): JSX.Element => {
+    //console.log('HIIIIII + ', slotRule);
```
rm

KevinTCoughlin commented on apps/theming-designer/src/components/Palette.tsx:
```diff
@@ -0,0 +1,101 @@
+    return (
+      <div key={slotRule.name} className="ms-themer-slot">
+        {this._colorSquareSwatchWidget(slotRule)}
+        <div>
```
superfluous DIVs can be removed as they aren't providing any value.

KevinTCoughlin commented on apps/theming-designer/src/components/Samples.tsx:
```diff
@@ -0,0 +1,132 @@
+            <SamplesCardHeader label="ACTION BUTTONS" />
+            <Stack horizontal gap={5}>
+              <ActionButton
+                data-automation-id="test"
```
Consider cleaning up the attributes / props copy pasted throughout as it'll be difficult to remove later. One example being `data-automation-id` which I doubt is used.

KevinTCoughlin commented on apps/theming-designer/src/components/ThemeDesignerColorPicker.tsx:
```diff
@@ -0,0 +1,22 @@
+const styles: any = stylesImport;
+
+export const ThemeDesignerColorPicker = (props: any) => {
+  const handleChange = (event: any, newVal: string | undefined) => {
```
Consider adding types for `any` throughout... becomes more difficult as the code grows.

ecraig12345 commented on apps/theming-designer/tsconfig.json:
```diff
@@ -0,0 +1,18 @@
+    "experimentalDecorators": true,
+    "forceConsistentCasingInFileNames": true,
+    "moduleResolution": "node",
+    "preserveConstEnums": true,
```
Can you add the following? These promote good practices and keep your project consistent with the others. (I'll also add this to the project template.)
```
"noImplicitAny": true,
"strictNullChecks": true,
```

ecraig12345 commented:
Re: @KevinTCoughlin's comments about `any`, it looks like your new project is missing a tslint.json. You should add one at the project root with the following contents:
```
{
  "extends": ["@uifabric/tslint-rules"],
  "rules": {}
}
```

aneeshack4 commented:
> Re: @KevinTCoughlin's comments about `any`, it looks like your new project is missing a tslint.json. You should add one at the project root with the following contents:
> 
> ```
> {
>   "extends": ["@uifabric/tslint-rules"],
>   "rules": {}
> }
> ```

added a ts lint!

KevinTCoughlin commented:
@aneeshack4 please ping me if/when you need this reviewed to merge

aneeshack4 commented:
@KevinTCoughlin I was mostly using this PR to start getting feedback early - I don't think I'll merge anytime soon because I'm still building a bunch of other things locally. I'll likely just make another PR with changes addressing these comments & the additions so I'll close this one for now.
